adr/0002-keycloak-gitlab.md

```diff
@@ -0,0 +1,165 @@
+
+### Provision Users
+Keycloak will provision users and manage user roles, groups, and permissions. This ensures user access is managed centrally and consistently across all applications and services.
+
+This may be done automatically on the first SSO login by the user or automated (Keycloak data periodically/realtime pushed to GitLab).
+
+Note: Supporting pushed users is helpful to GitLab users, because this allows other users within the GitLab environment to begin interacting with the new user before their first login.
+
+#### Automatic provisioning on first SSO login
+**Pros:**
+- Common standard SSO integration method
+- The new GitLab user account is created on demand, no provisioning delay for the new user to login.
+
+**Cons:**
+- New user accounts don't exist in GitLab before the new user's first login. This prevents other users in GitLab from interacting with the new user (assigning tasks, @ mentions, etc) before their first login.
+
+#### SCIM Integration
+
+Note: Using SCIM integration in GitLab requires using SAML for authentication. See [Configure SCIM for self-managed GitLab instances](https://docs.gitlab.com/ee/administration/settings/scim_setup.html) documentation.
+
+**Pros:**
+- Easy to Mix Systems: SCIM is a common framework for sharing user info between different programs or online services, making them work smoothly.
+- Saves Time: It handles adding, updating, and removing user accounts by itself, which means less work for people managing these tasks.
+- Fewer Mistakes: Since it's automated, there's less chance someone will type something wrong or forget to update important info.
+
+**Cons:**
+- Keycloak does not natively support SCIM. There are some open source plugins to add support, but they would need review.
+
+#### Manual Provisioning
+**Pros:**
+- Full control over the user provisioning process.
+- Can be implemented without additional development or integration efforts.
+
+**Cons:**
+- Time-consuming and not scalable.
+- Prone to human error, potentially leading to security and compliance issues.
+
+#### Custom Code
+**Pros:**
+- Flexibility: You can tailor the provisioning process to your exact needs, ensuring it fits your organization perfectly.
+- No exteranl depdency: You can do this with what GitLab already offers, no need for external OSS tools.
+
+**Cons:**
+- More Work: You'll have to write and maintain the code yourself, which can be a lot of work.
+
+### Deprovision Users
+Users need to be deprovisioned (GitLab user account disabled or deleted) promptly from GitLab when they lose access via Keycloak (Keycloak user account disabled/deleted or removed from a group that gives GitLab access).
+
+Note: it possible to authenticate to GitLab with Personal Access Tokens (PATs) and SSH keys, these bypass SSO because the user does not login with a browser. It is critical to promptly deactivate or delete users from GitLab to terminate access.
```

Comment:
```suggestion
Note: it is possible to authenticate to GitLab with Personal Access Tokens (PATs) and SSH keys, these bypass SSO because the user does not login with a browser. It is critical to promptly deactivate or delete users from GitLab to terminate access.
```

Reply:
👍